File: sdks/js/src/keys.ts
/** DocParse key management — generate, list, revoke, rotate, usage. */

import type { KeyInfo, UsageInfo } from "./types.js";
import type { DocParse } from "./client.js";

export class KeyManager {
  private client: DocParse;

  constructor(client: DocParse) {
    this.client = client;
  }

  /** Generate a new API key. Raw key is shown once. */
  async generate(label = "default", userId = ""): Promise<KeyInfo> {
    return this.client._call("POST", "/api/v1/keys/generate", [userId, label]);
  }

  /** List API keys for a user. */
  async list(userId = ""): Promise<any> {
    return this.client._call("POST", "/api/v1/keys/list", [userId]);
  }

  /** Revoke an API key. */
  async revoke(keyId: string, userId = ""): Promise<any> {
    return this.client._call("POST", "/api/v1/keys/revoke", [keyId, userId]);
  }

  /** Rotate a key — generates new, revokes old, preserves tier. */
  async rotate(keyId: string, userId = ""): Promise<KeyInfo> {
    return this.client._call("POST", "/api/v1/keys/rotate", [keyId, userId]);
  }

  /** Get usage statistics for a key. */
  async usage(keyId: string, userId = ""): Promise<UsageInfo> {
    return this.client._call("POST", "/api/v1/keys/usage", [keyId, userId]);
  }
}

File: sdks/go/keys.go
package docparse

import (
	"context"
	"encoding/json"
	"fmt"
)

// KeyManager provides API key management methods.
type KeyManager struct {
	client *Client
}

// Generate creates a new API key. The raw key is shown once.
func (km *KeyManager) Generate(ctx context.Context, label, userID string) (*KeyInfo, error) {
	data, err := km.client.call(ctx, "POST", "/api/v1/keys/generate", []string{userID, label})
	if err != nil {
		return nil, err
	}
	var result KeyInfo
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("unmarshal key info: %w", err)
	}
	return &result, nil
}

// List returns API keys for a user.
func (km *KeyManager) List(ctx context.Context, userID string) (json.RawMessage, error) {
	data, err := km.client.call(ctx, "POST", "/api/v1/keys/list", []string{userID})
	if err != nil {
		return nil, err
	}
	return json.RawMessage(data), nil
}

// Revoke revokes an API key.
func (km *KeyManager) Revoke(ctx context.Context, keyID, userID string) error {
	_, err := km.client.call(ctx, "POST", "/api/v1/keys/revoke", []string{keyID, userID})
	return err
}

// Rotate generates a new key and revokes the old one, preserving tier.
func (km *KeyManager) Rotate(ctx context.Context, keyID, userID string) (*KeyInfo, error) {
	data, err := km.client.call(ctx, "POST", "/api/v1/keys/rotate", []string{keyID, userID})
	if err != nil {
		return nil, err
	}
	var result KeyInfo
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("unmarshal key info: %w", err)
	}
	return &result, nil
}

// Usage returns usage statistics for a key.
func (km *KeyManager) Usage(ctx context.Context, keyID, userID string) (*UsageInfo, error) {
	data, err := km.client.call(ctx, "POST", "/api/v1/keys/usage", []string{keyID, userID})
	if err != nil {
		return nil, err
	}
	var result UsageInfo
	if err := json.Unmarshal(data, &result); err != nil {
		return nil, fmt.Errorf("unmarshal usage info: %w", err)
	}
	return &result, nil
}
